diff -rupw libexo/src/xfce-git-treeview/exo/exo-tree-view.c lxdesktop/src/gtk/exo/exo-tree-view.c
--- libexo/src/xfce-git-treeview/exo/exo-tree-view.c	2012-04-30 14:58:10.000000000 +0200
+++ lxdesktop/src/gtk/exo/exo-tree-view.c	2012-05-19 05:28:49.759177313 +0200
@@ -21,12 +21,8 @@
 #include <config.h>
 #endif
 
-#include <exo/exo-config.h>
-#include <exo/exo-private.h>
-#include <exo/exo-string.h>
-#include <exo/exo-tree-view.h>
-#include <exo/exo-utils.h>
-#include <exo/exo-alias.h>
+// AXL_CHANGES:
+#include "exo-lxde.h"
 
 /**
  * SECTION: exo-tree-view
@@ -316,7 +357,7 @@ exo_tree_view_button_press_event (GtkWid
       /* if no custom select function is set, we simply use exo_noop_false here,
        * to tell the tree view that it may not alter the selection.
        */
-      if (G_LIKELY (selection->user_func == NULL))
+      if (G_LIKELY (gtk_tree_selection_get_select_function (selection) == NULL))
         gtk_tree_selection_set_select_function (selection, (GtkTreeSelectionFunc) exo_noop_false, NULL, NULL);
       else
         selected_paths = gtk_tree_selection_get_selected_rows (selection, NULL);
@@ -360,12 +401,16 @@ exo_tree_view_button_press_event (GtkWid
   /* call the parent's button press handler */
   result = (*GTK_WIDGET_CLASS (exo_tree_view_parent_class)->button_press_event) (widget, event);
 
+  // LXDE_CHANGES:
+  if( treat_as_blank )
+    gtk_tree_selection_unselect_all( selection );
+
   /* restore previous selection if the path is still selected */
   if (event->type == GDK_BUTTON_PRESS && (event->state & gtk_accelerator_get_default_mod_mask ()) == 0
       && path != NULL && gtk_tree_selection_path_is_selected (selection, path))
     {
       /* check if we have to restore paths */
-      if (G_LIKELY (selection->user_func != (GtkTreeSelectionFunc) exo_noop_false))
+      if (G_LIKELY (gtk_tree_selection_get_select_function (selection) != (GtkTreeSelectionFunc) exo_noop_false))
         {
           /* select all previously selected paths */
           for (lp = selected_paths; lp != NULL; lp = lp->next)
@@ -373,15 +418,17 @@ exo_tree_view_button_press_event (GtkWid
         }
     }
 
+  // NOTE_axl: http://git.xfce.org/xfce/exo/commit/?id=cd80ae7747aa87ab6d72f08386b0fd4bea7f05b3
   /* see bug http://bugzilla.xfce.org/show_bug.cgi?id=6230 for more information */
-  if (G_LIKELY (selection->user_func == (GtkTreeSelectionFunc) exo_noop_false))
+  if (G_LIKELY (gtk_tree_selection_get_select_function (selection) == (GtkTreeSelectionFunc) exo_noop_false))
     {
       /* just reset the select function (previously set to exo_noop_false),
        * there's no clean way to do this, so what the heck.
        */
-      selection->user_func = NULL;
+      gtk_tree_selection_set_select_function (selection, (GtkTreeSelectionFunc) NULL, NULL, NULL);
     }
 
+  
   /* release the path (if any) */
   if (G_LIKELY (path != NULL))
     gtk_tree_path_free (path);
@@ -518,7 +578,7 @@ exo_tree_view_motion_notify_event (GtkWi
                   /* setup the hand cursor to indicate that the row at the pointer can be activated with a single click */
                   cursor = gdk_cursor_new (GDK_HAND2);
                   gdk_window_set_cursor (event->window, cursor);
-                  gdk_cursor_unref (cursor);
+                  g_object_unref (cursor);
                 }
               else
                 {
@@ -575,7 +635,7 @@ exo_tree_view_leave_notify_event (GtkWid
     }
 
   /* reset the cursor for the tree view internal window */
-  if (GTK_WIDGET_REALIZED (tree_view))
+  if (gtk_widget_get_realized ((GtkWidget*) tree_view))
     gdk_window_set_cursor (gtk_tree_view_get_bin_window (GTK_TREE_VIEW (tree_view)), NULL);
 
   /* the next button-release-event should not activate */
@@ -621,7 +681,7 @@ exo_tree_view_move_cursor (GtkTreeView
     }
 
   /* reset the cursor for the tree view internal window */
-  if (GTK_WIDGET_REALIZED (tree_view))
+  if (gtk_widget_get_realized ((GtkWidget*) tree_view))
     gdk_window_set_cursor (gtk_tree_view_get_bin_window (GTK_TREE_VIEW (tree_view)), NULL);
 
   /* call the parent's handler */
@@ -646,7 +706,7 @@ exo_tree_view_single_click_timeout (gpoi
   GDK_THREADS_ENTER ();
 
   /* verify that we are in single-click mode, have focus and a hover path */
-  if (GTK_WIDGET_HAS_FOCUS (tree_view) && tree_view->priv->single_click && tree_view->priv->hover_path != NULL)
+  if (gtk_widget_has_focus ((GtkWidget*) tree_view) && tree_view->priv->single_click && tree_view->priv->hover_path != NULL)
     {
       /* transform the hover_path to a tree iterator */
       model = gtk_tree_view_get_model (GTK_TREE_VIEW (tree_view));
